### PR TITLE
Support multi-digit macro arguments in angle brackets

### PR DESCRIPTION
**Pull request by Rangi42**
This allows access to arguments past `\9` without using `shift`.

### CODE REVIEW

**Comment by Rangi42** [comment]
There's no open issue for this feature but I remember it being suggested before. (Full expressions inside parentheses wouldn't be practical because that's parsing, not lexing.)

**Comment by Rangi42** [comment]
This also removes the out-of-date documentation lines about macros having max 256 arguments.

**Comment by ISSOtm** [comment]
While I certainly don't have any problems with the code change (except maybe how many forward declarations are needed? :S), I'm less sure about the syntax change.

This syntax couldn't be `\{42}`, for example, because that's already an escape used to remove `{`'s special meaning. But then, giving `\(` special meaning means that **we won't be able to give `(` any special meaning**. There's nothing wrong with the change *per se*, but I want us to be sure that we won't regret this decision later.

**Comment by Rangi42** [comment]
Agreed about the forward declarations; I'll see if there's a rearrangement that reduces their overall count.

You're right about parentheses: #704 proposes keeping track of parentheses nesting in macro arguments and allowing commas inside them, so e.g. `mac fn(1, 2)` would pass `fn(1, 2)` as `\1`. To disable that the parentheses would have to be escapable, like `mac fn\(1, 2)`.

Alternate syntaxes: brackets `\[10]`, angle brackets `\<10>`, or various paired punctuation (`\$10$`, `\^10^`, `\:10:`, etc). (Not single quotes `\'10'`, that would conflict with potentially escaping `\'` within `''` for character literals or some kind of special string.) Brackets might be getting used for inline fragments and/or structs, and while neither of those uses currently demands a `\[` escape, it's plausible that we'll end up with one. I don't *think* `\<` would ever be necessary though; it's just for the `<` and `<<` operators. How about it?

**Comment by ISSOtm** [comment]
`\$` sounds like hex, so not that one; `\^` looks weird imo. That seems to leave `\:` and `\<`, so I guess `\<`. It should be possible for us to be bitten regardless, but it seems unlikely, so the risk is imo acceptable.

File: src/asm/lexer.c [hunk 1]
```diff
@@ -717,21 +717,94 @@ static void freeExpansion(struct Expansion *expansion)
...
+			fatalerror("Bracketed symbol \"%s\" is not numeric\n", symName);
+
+		num = sym_GetConstantSymValue(sym);
+	} else {
+		empty = true;
+	}
+
+	c = peek();
+	if (c != '>')
+		fatalerror("Invalid character in bracketed macro argument %s\n", printChar(c));
```
Comment (ISSOtm) [inline src/asm/lexer.c:775]: Do these need to be fatal? An invalid char could cause the bracketing to implicitly stop, and an empty/0 argument to be expanded to nothing.

File: src/asm/lexer.c [hunk 2]
```diff
@@ -717,21 +717,94 @@ static void freeExpansion(struct Expansion *expansion)
...
+		size_t i = 0;
+
+		for (; continuesIdentifier(c); c = peek()) {
+			if (i < sizeof(symName))
+				symName[i++] = c;
+			shiftChar();
+		}
+
+		if (i == sizeof(symName)) {
+			warning(WARNING_LONG_STR, "Symbol name too long\n");
```
Comment (ISSOtm) [inline src/asm/lexer.c:756]: "... in bracketed macro arg"?

**Review by ISSOtm — CHANGES_REQUESTED** [review]

File: src/asm/lexer.c (hunk 1 above)
Comment (Rangi42) [reply src/asm/lexer.c:775]: I made them fatal to match the fatal "`Invalid macro argument '\0'`" and "`Macro argument '\%c' not defined`" errors. IMO either they should all be fatal, or all become empty expansions.

Comment (ISSOtm) [reply src/asm/lexer.c:775]: I'd prefer the latter, though it can be a separate issue then.

Comment (Rangi42) [reply src/asm/lexer.c:775]: Non-fatal errors for them all would match interpolations, where `{undefinedsymbol}` and `{SomeLabel}` are merely `error`s. I can do that as a separate PR/commit.

Comment (ISSOtm) [reply src/asm/lexer.c:775]: See #850 :P

**Comment by Rangi42** [comment]
(Angle brackets are kind of appropriate for being related to macro args. Some older assemblers used them to surround argument *values* that needed actual commas, e.g. `mac <1,2>,3` would have `\1` be `1,2` and `\2` be `3`.)

(Also, [this assembler](https://acorn.huininga.nl/pub/projects/CiscOS/_emulators/EASy68Ksource/EASy68K_Help/Directives/macro.htm) allowed 36 arguments, `\0`-`\9` and then `\A`-`\Z`. Being case-sensitive, `\N` wouldn't even conflict with `\n`. :P Of course that's irrelevant now with `\<36>` and above.)

**Review by ISSOtm — APPROVED** [review]

**Pull request merged**